### PR TITLE
More PropertyGrid related cleanup

### PR DESCRIPTION
**Pull request by JeremyKuhne**
- Reduce nesting by introducing FreezePaintScope
- Further Reduce nesting in SelectedObjects setter and tweak naming and comments.
- Clean up obtuse EventsTab.GetProperties

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/5441)

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
Closed and reopened a few times to get the CI working correctly.

File: src/System.Windows.Forms/src/System/Windows/Forms/Design/EventsTab.cs [hunk 1]
```diff
@@ -16,72 +15,64 @@ public class EventsTab : PropertyTab
...
+            return TypeDescriptor.GetDefaultEvent(obj) is EventDescriptor defaultEvent
+                ? eventPropertyService.GetEventProperty(defaultEvent)
+                : null;
         }
 
-        private IEventBindingService? GetEventPropertyService(object obj, ITypeDescriptorContext? context)
+        /// <summary>
+        ///  Returns the most relevant <see cref="IEventBindingService"/>.
+        ///
+        ///  This looks first at the current <see cref="IDesignerHost"/>, then the <paramref name="object"/>'s
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Design/EventsTab.cs:55]: nit: `<para>` tag

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs [hunk 2]
```diff
@@ -2674,181 +2613,174 @@ private void OnLayoutInternal(bool dividerOnly)
...
-                    if (hcRequestedHeight == 0 || (dcRequestedHeight + hcRequestedHeight) < maxSpace)
+                    hcOptHeight = _hotCommands.GetOptimalHeight(Size.Width - s_cyDivider);
+                    if (_hotCommands.UserSized)
                     {
-                        // Full size.
-                        height = Math.Min(dcRequestedHeight, maxSpace);
+                        hcRequestedHeight = _hotCommands.Size.Height;
                     }
-                    else if (hcRequestedHeight > 0 && hcRequestedHeight < maxSpace)
+                    else if (_hotCommandsSizeRatio != -1)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs:2699]: Is it worth introducing a const for `-1`

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs [hunk 3]
```diff
@@ -3315,144 +3236,139 @@ protected override void OnVisibleChanged(EventArgs e)
...
-                    {
-                        break;
-                    }
+                return base.ProcessDialogKey(keyData);
+            }
 
-                    // Are we going forward?
-                    if ((keyData & Keys.Shift) != 0)
+            // Are we going forward?
+            if ((keyData & Keys.Shift) != 0)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs:3247]: Would `Enum.HasFlag` work here?

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs [hunk 4]
```diff
@@ -3981,194 +3890,187 @@ private void SetupToolbar(bool fullRebuild)
...
+                // Add the viewType buttons and a separator.
+                for (i = 0; i < _viewSortButtons.Length; i++)
+                {
+                    buttonList.Add(_viewSortButtons[i]);
+                }
+            }
+            else
+            {
+                // Clear all the items from the toolStrip and image list after the first two.
+                int items = buttonList.Count;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs:3975]: `count` might be a better name

**Review by Tanya-Solyanik — APPROVED** [review]
I added some nits

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs:2699]: For -1 I don't think so. It is a pretty standard uninitialized value. Other values I probably would.

**Review by RussKie — APPROVED** [review]
LGTM

File: src/System.Windows.Forms/src/System/Windows/Forms/Design/EventsTab.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/Design/EventsTab.cs:55]: Ugh. I hate the `<para>` tag.

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/Design/EventsTab.cs:55]: remarks?

Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/Design/EventsTab.cs:55]: I did change it and moved the second one to remarks, which complains about the <para> if it isn't present.

**Comment by RussKie** [comment]
Please rebase you work on to the main. vNext branch is superficial.

**Pull request merged**